Neos.Flow/Tests/Unit/Mvc/Controller/ActionControllerTest.php

```diff
@@ -246,6 +248,66 @@ public function processRequestInjectsSettingsToView()
+            [['text/html', 'application/json'], 'application/json', 'application/json'],
+            [['text/html'], 'text/html, application/xhtml+xml, application/xml;q=0.9, */*;q=0.8', 'text/html'],
+            [['application/json', 'application/xml'], 'text/html, application/json;q=0.7, application/xml;q=0.9', 'application/xml'],
+        ];
+    }
+
+    /**
+     * @test
+     * @dataProvider supportedAndRequestedMediaTypes
+     */
+    public function processRequestSetsNegotiatedContentTypeOnResponse($supportedMediaTypes, $acceptHeader, $expected)
+    {
+        $this->actionController = $this->getAccessibleMock(ActionController::class, ['resolveActionMethodName', 'initializeActionMethodArguments', 'initializeActionMethodValidators', 'resolveView', 'callActionMethod']);
+
+        $this->inject($this->actionController, 'objectManager', $this->mockObjectManager);
+
+        $mockMvcPropertyMappingConfigurationService = $this->createMock(Mvc\Controller\MvcPropertyMappingConfigurationService::class);
+        $this->inject($this->actionController, 'mvcPropertyMappingConfigurationService', $mockMvcPropertyMappingConfigurationService);
+
+        $mockHttpRequest = $this->getMockBuilder(ServerRequestInterface::class)->disableOriginalConstructor()->getMock();
+        $mockHttpRequest->method('getHeaderLine')->with('Accept')->willReturn($acceptHeader);
+        $this->mockRequest->method('getHttpRequest')->willReturn($mockHttpRequest);
+
+        $mockResponse = new Mvc\ActionResponse;
+        $this->inject($this->actionController, 'supportedMediaTypes', $supportedMediaTypes);
+
+        $this->actionController->processRequest($this->mockRequest, $mockResponse);
+        self::assertSame($expected, $mockResponse->getContentType());
+    }
+
+    /**
+     * @test
+     * @dataProvider supportedAndRequestedMediaTypes
+     */
+    public function processRequestUsesContentTypeFromActionResponse($supportedMediaTypes, $acceptHeader, $expected)
+    {
+        $this->actionController = $this->getAccessibleMock(ActionController::class, ['resolveActionMethodName', 'initializeActionMethodArguments', 'initializeActionMethodValidators', 'resolveView', 'callActionMethod']);
+
+        $this->inject($this->actionController, 'objectManager', $this->mockObjectManager);
+
+        $mockMvcPropertyMappingConfigurationService = $this->createMock(Mvc\Controller\MvcPropertyMappingConfigurationService::class);
+        $this->inject($this->actionController, 'mvcPropertyMappingConfigurationService', $mockMvcPropertyMappingConfigurationService);
+
+        $mockHttpRequest = $this->getMockBuilder(ServerRequestInterface::class)->disableOriginalConstructor()->getMock();
+        $mockHttpRequest->method('getHeaderLine')->with('Accept')->willReturn('application/xml');
+        $this->mockRequest->method('getHttpRequest')->willReturn($mockHttpRequest);
+
+        $mockResponse = new Mvc\ActionResponse;
+        $mockResponse->setContentType('application/json');
```

Comment:
This test should show the actual regression that got introduced with #2005 
It also shows that the action can completely ignore the client accepted media-types - maybe we could at least add a feature at some point that checks the content-type of the ActionResponse against the accept header and logs a warning when they mismatch.